WifiDirectTest/app/src/main/java/test/microsoft/com/wifidirecttest/WifiDirectService.java

```diff
@@ -0,0 +1,411 @@
+import static android.net.wifi.p2p.WifiP2pManager.WIFI_P2P_PEERS_CHANGED_ACTION;
+import static android.net.wifi.p2p.WifiP2pManager.WIFI_P2P_STATE_CHANGED_ACTION;
+import static android.net.wifi.p2p.WifiP2pManager.WIFI_P2P_THIS_DEVICE_CHANGED_ACTION;
+
+/**
+ * Created by juksilve on 17.2.2015.
+ */
+public class WifiDirectService extends Service {
+
+    static final public String DSS_WIFIDIRECT_VALUES = "test.microsoft.com.wifidirecttest.DSS_WIFIDIRECT_VALUES";
+    static final public String DSS_WIFIDIRECT_MESSAGE = "test.microsoft.com.wifidirecttest.DSS_WIFIDIRECT_MESSAGE";
+
+
+    private static final String TAG = MainActivity.class.getPackage().getName();
+
+    private static final String SERVICE_TYPE = "_p2p._tcp";
+
+    LocalBroadcastManager broadcaster;
+    MyTextSpeech mySpeech = null;
+
+    private WifiP2pManager p2p;
+    private WifiP2pManager.Channel channel;
+    private volatile ServerSocket wifiSocket = null;
+
+    //can we make this dynamic ?
+    private static final int PORT = 45678;
+
+    private BroadcastReceiver receiver;
+    private IntentFilter filter;
+
+    private WifiP2pManager.PeerListListener peerListListener;
+    private WifiP2pManager.DnsSdServiceResponseListener serviceListener;
+    private WifiP2pManager.DnsSdTxtRecordListener txtListener;
+
+    private WifiP2pDnsSdServiceInfo serviceInfo;
+
+    private final IBinder mBinder = new LocalBinder();
+    public class LocalBinder extends Binder {
+        WifiDirectService getService() {
+            return WifiDirectService.this;
+        }
+    }
+    @Override
+    public IBinder onBind(Intent intent) {
+        return mBinder;
+    }
+
+    public int onStartCommand(Intent intent, int flags, int startId) {
+        return START_STICKY;
+    }
+
+    @Override
+    public void onCreate() {
+        mySpeech = new MyTextSpeech(this);
+
+        broadcaster = LocalBroadcastManager.getInstance(this);
+
+
+        p2p = (WifiP2pManager) getSystemService(WIFI_P2P_SERVICE);
+        if (p2p == null) {
+            mySpeech.speak("This device does not support Wi-Fi Direct");
+            sendResult("This device does not support Wi-Fi Direct");
+        } else {
+            channel = p2p.initialize(this, getMainLooper(), null);
+
+            receiver = new PeerReceiver();
+            filter = new IntentFilter();
+            filter.addAction(WIFI_P2P_STATE_CHANGED_ACTION);
+            filter.addAction(WIFI_P2P_PEERS_CHANGED_ACTION);
+            filter.addAction(WIFI_P2P_CONNECTION_CHANGED_ACTION);
+            filter.addAction(WIFI_P2P_THIS_DEVICE_CHANGED_ACTION);
+            registerReceiver(receiver, filter);
+
+            openWifiSocket();
+
+            peerListListener = new WifiP2pManager.PeerListListener() {
+
+                public void onPeersAvailable(WifiP2pDeviceList peers) {
+                    sendResult("Discovered peers:");
+
+                    final WifiP2pDeviceList pers = peers;
```

Comment:
I typically follow the convention of finalX for this case. E.g. finalPeers rather than pers. That way it's easier to tell from reading the code what is going on.